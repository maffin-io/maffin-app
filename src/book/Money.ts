import * as currencies from '@dinero.js/currencies';
import type { Currency as DineroCurrency } from 'dinero.js';
import * as djs from 'dinero.js';

import { toAmountWithScale, moneyToString } from '@/helpers/number';

export default class Money {
  private _raw: djs.Dinero<number>;

  /**
   * A wrapper class for dinero object to reduce verbosity and overhead. The constructor
   * checks for how many decimals the amount has and calculates the scale automatically.
   *
   * @param amount - The amount to be considered in the money object. Can be a float
   * @param currency - The currency
   * @param scale - Number representing custom scale. Default is 0, if passed differently, that one
   * is taken into account
   */
  constructor(amount: number, currency: string, scale?: number) {
    const dineroCurrency = toCurrency(currency);

    if (scale === undefined) {
      const { amount: newAmount, scale: newScale } = toAmountWithScale(amount);
      this._raw = djs.dinero({
        currency: dineroCurrency,
        amount: newAmount,
        scale: newScale,
      });
    } else {
      this._raw = djs.dinero({
        currency: dineroCurrency,
        amount,
        scale,
      });
    }
  }

  get currency(): string {
    return djs.toSnapshot(this._raw).currency.code;
  }

  get raw(): djs.Dinero<number> {
    return this._raw;
  }

  /**
   * Represents the money object as a string with the currency standardizing
   * the scale to 2 as a default. If the scale is passed then it uses that.
   */
  toString(scale = 2): string {
    return djs.toDecimal(
      djs.transformScale(this._raw, scale),
      ({ value, currency }) => `${value} ${currency.code}`,
    );
  }

  /**
   * Represents the money object as a string with the currency standardizing
   * the scale to 2 as a default. If the scale is passed then it uses that.
   *
   * The result is a localized string with the currency as a symbol
   */
  format(scale = 2): string {
    return djs.toDecimal(
      djs.transformScale(this._raw, scale),
      ({ value, currency }) => moneyToString(Number(value), currency.code),
    );
  }

  toNumber(): number {
    return parseFloat(djs.toDecimal(this._raw));
  }

  isNegative(): boolean {
    return djs.isNegative(this._raw);
  }

  isZero(): boolean {
    return djs.isZero(this._raw);
  }

  /**
   * Adds b to the current object and returns a new one with the result
   *
   * @param b - Money to be added
   * @returns - Money with result
   */
  add(b: Money): Money {
    const result = djs.toSnapshot(djs.add(b.raw, this.raw));
    return new Money(result.amount, this.currency, result.scale);
  }

  /**
   * Subtracts b to the current object and returns a new one with the result
   *
   * @param b - Money to be subtracted
   * @returns - Money with result
   */
  subtract(b: Money): Money {
    const result = djs.toSnapshot(djs.subtract(this.raw, b.raw));
    return new Money(result.amount, this.currency, result.scale);
  }

  /**
   * Multiplies b to the current object and returns a new one with the result
   * Initially, we try to multiply without transforming b but sometimes dinerojs
   * returns that it is an invalid amount (I have no idea why). If it fails, then
   * we transform b to `amountWithScale` and multiply with that.
   *
   * @param b - amount to be multiplied to in scale 0
   * @returns - Money with result
   */
  multiply(b: number): Money {
    let result;

    try {
      result = djs.multiply(this.raw, b);
    } catch {
      result = djs.multiply(this.raw, toAmountWithScale(b));
    }

    const snapshot = djs.toSnapshot(result);
    return new Money(snapshot.amount, this.currency, snapshot.scale);
  }

  /**
   * Checks for equality with b by comparing amount and currency
   *
   * @returns - boolean
   */
  equals(b: Money): boolean {
    return djs.equal(this._raw, b._raw);
  }

  /**
   * Converts a Money object to the currency specified
   *
   * @param to - Money object containing the rate to
   */
  convert(
    to: string,
    rate: number,
  ): Money {
    const rates = {
      [to]: {
        ...toAmountWithScale(rate),
      },
    };
    const currency = toCurrency(to);

    try {
      // @ts-ignore
      const result = djs.toSnapshot(djs.convert(this.raw, currency, rates));
      return new Money(result.amount, currency.code, result.scale);
    } catch (e) {
      throw new Error(`Error converting ${this.format()} to ${to} with rates ${JSON.stringify(rates)}: ${e}`);
    }
  }
}

function toCurrency(currency: string): DineroCurrency<number> {
  // @ts-ignore
  let dineroCurrency = currencies[currency];
  if (!dineroCurrency) {
    dineroCurrency = {
      code: currency,
      base: 10,
      exponent: 2,
    };
  }
  return dineroCurrency;
}
