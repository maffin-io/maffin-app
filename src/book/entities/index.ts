export { default as Account } from './Account';
export { default as BankConfig } from './BankConfig';
export { default as Commodity } from './Commodity';
export { default as Price } from './Price';
export { default as Split } from './Split';
export { default as Transaction } from './Transaction';
export { default as Book } from './Book';
export { default as BaseEntity } from './BaseEntity';
