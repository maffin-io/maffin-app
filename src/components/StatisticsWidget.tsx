import React from 'react';

export type StatisticsWidgetProps = {
  title: string,
  stats: string,
  description: string,
  className?: string,
  statsTextClass?: string,
};

export default function StatisticsWidget({
  title,
  stats,
  description,
  className = '',
  statsTextClass = '',
}: StatisticsWidgetProps): JSX.Element {
  return (
    <div className={`${className} bg-gunmetal-700 rounded-sm p-6`}>
      <p>
        {title}
      </p>
      <p className={`${statsTextClass} text-2xl font-semibold my-3`}>{stats}</p>

      <span className="text-sm">
        {description}
      </span>
    </div>
  );
}
