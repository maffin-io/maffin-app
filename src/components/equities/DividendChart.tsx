import React from 'react';
import dynamic from 'next/dynamic';

import { toFixed } from '@/helpers/number';
import type { InvestmentAccount } from '@/book/models';
import { currencyToSymbol } from '@/book/helpers';

// apexcharts import references window so we need this
// https://stackoverflow.com/questions/68596778
const Chart = dynamic(() => import('react-apexcharts'), { ssr: false });

const MONTHS: { [key:number]:string; } = {
  0: 'Jan',
  1: 'Feb',
  2: 'Mar',
  3: 'Apr',
  4: 'May',
  5: 'Jun',
  6: 'Jul',
  7: 'Aug',
  8: 'Sep',
  9: 'Oct',
  10: 'Nov',
  11: 'Dec',
};

export type DividendChartProps = {
  investments: InvestmentAccount[],
};

export default function DividendChart({
  investments,
}: DividendChartProps): JSX.Element {
  if (!investments.length) {
    return (
      <span>Loading...</span>
    );
  }

  const currency = investments[0].mainCurrency;
  const series = buildSeries(investments);

  return (
    <div className="bg-gunmetal-700 rounded-sm m-6 p-4">
      <span className="text-lg">
        {`
          Dividends in
          ${currencyToSymbol(currency)}
        `}
      </span>

      <div className="grid grid-cols-12">
        <div className="col-span-4">
          <div id="chart-year" className="text-slate-300">
            {
              (typeof window !== 'undefined')
              // @ts-ignore
              && <Chart options={OPTIONS_YEARLY} series={[{ data: series }]} type="bar" width="100%" height={400} />
            }
          </div>
        </div>
        <div className="col-span-8">
          <div id="chart-monthly">
            {
              (typeof window !== 'undefined')
              && (
                // @ts-ignore
                <Chart
                  options={OPTIONS_MONTHLY}
                  series={[]}
                  type="bar"
                  className="apex-charts"
                  // https://stackoverflow.com/questions/75103994
                  width="100%"
                  height={400}
                />
              )
            }
          </div>
        </div>
      </div>
    </div>
  );
}

function buildSeries(investments: InvestmentAccount[]) {
  const perYear: {
    [year:string]: {
      [month:string]: {
        amount: number,
        ticker: string,
      },
    },
  } = {};
  investments.forEach(investment => {
    investment.dividends.forEach(dividend => {
      const month = MONTHS[dividend.when.month - 1];
      const year = dividend.when.year.toString();

      if (!(year in perYear)) {
        perYear[year] = {};
      }

      const amount = dividend.amountInCurrency.toNumber();

      if (!(month in perYear[year])) {
        // @ts-ignore
        perYear[year][month] = [{
          amount,
          ticker: investment.account.name,
        }];
      } else {
        // @ts-ignore
        perYear[year][month].push({
          amount,
          ticker: investment.account.name,
        });
      }
    });
  });

  // @ts-ignore
  const series = [];
  Object.entries(perYear).forEach(([year, perMonth]) => {
    let yearTotal = 0;
    Object.values(perMonth).forEach(monthlyDividend => {
      // @ts-ignore
      monthlyDividend.forEach(dividend => {
        yearTotal += dividend.amount;
      });
    });

    const yearSeries = {
      x: year,
      y: yearTotal,
      dividends: perMonth,
    };
    series.push(yearSeries);
  });

  // @ts-ignore
  return series;
}

// @ts-ignore
function updateMonthlyChart(yearlyChart, destChartIDToUpdate) {
  const seriesIndex = 0;

  const selectedPoint = yearlyChart.w.globals.selectedDataPoints[0][0];
  if (selectedPoint === undefined) {
    return [];
  }
  const selectedSeries = yearlyChart.w.config.series[seriesIndex].data[selectedPoint];

  const tickerSeries = {};

  Object.entries(MONTHS).forEach(([monthNumber, month]) => {
    if (month in selectedSeries.dividends) {
      // @ts-ignore
      selectedSeries.dividends[month].forEach(dividend => {
        if (!(dividend.ticker in tickerSeries)) {
          // @ts-ignore
          tickerSeries[dividend.ticker] = new Array(12).fill(0);
        }
        // @ts-ignore
        tickerSeries[dividend.ticker][monthNumber] = dividend.amount;
      });
    }
  });

  // @ts-ignore
  const series = [];
  Object.entries(tickerSeries).forEach(([ticker, monthlyDividends]) => {
    series.push({
      name: ticker,
      data: monthlyDividends,
    });
  });

  return ApexCharts.exec(destChartIDToUpdate, 'updateOptions', {
    // @ts-ignore
    series,
  });
}

const OPTIONS_YEARLY = {
  grid: {
    show: false,
  },
  chart: {
    id: 'barYear',
    width: '100%',
    foreColor: '#94A3B8',
    events: {
      // @ts-ignore
      dataPointSelection: (e, chart) => {
        updateMonthlyChart(chart, 'barMonthly');
      },
      // @ts-ignore
      mounted: (chart) => {
        const seriesIndex = 0;
        if (chart.w.globals.selectedDataPoints.length === 0) {
          // Display the last year by default
          chart.toggleDataPointSelection(
            seriesIndex,
            chart.w.config.series[seriesIndex].data.length - 1,
          );
        }

        updateMonthlyChart(chart, 'barMonthly');
      },
    },
    toolbar: {
      show: false,
    },
  },
  plotOptions: {
    bar: {
      horizontal: true,
      barHeight: '55%',
    },
  },
  dataLabels: {
    enabled: false,
  },
  states: {
    active: {
      allowMultipleDataPointsSelection: false,
      filter: {
        type: 'lighten',
        value: 0.5,
      },
    },
  },
  tooltip: {
    fillSeriesColor: true,
    y: {
      title: {
        formatter: () => '',
      },
      formatter: (val: number) => `${toFixed(val)}`,
    },
    x: {
      show: false,
    },
    theme: 'dark',
  },
  yaxis: {
    labels: {
      show: true,
    },
  },
  legend: {
    show: false,
  },
};

const OPTIONS_MONTHLY = {
  grid: {
    borderColor: '#777f85',
  },
  chart: {
    id: 'barMonthly',
    width: '100%',
    stacked: true,
    foreColor: '#94A3B8',
    toolbar: {
      show: false,
    },
  },
  yaxis: [
    {
      labels: {
        formatter: (val: number) => `${toFixed(val)}`,
      },
    },
  ],
  xaxis: {
    categories: Object.values(MONTHS),
    axisBorder: {
      show: false,
    },
  },
  plotOptions: {
    bar: {
      horizontal: false,
      columnWidth: '55%',
      endingShape: 'rounded',
    },
  },
  dataLabels: {
    enabled: false,
  },
  tooltip: {
    fillSeriesColor: true,
    y: {
      formatter: (val: number) => `${toFixed(val)}`,
    },
    x: {
      show: false,
    },
    theme: 'dark',
    intersect: true,
    inverseOrder: true,
    shared: false,
  },
  legend: {
    show: true,
  },
};
