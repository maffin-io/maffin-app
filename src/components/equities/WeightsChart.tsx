import React from 'react';
import dynamic from 'next/dynamic';
import { ApexOptions } from 'apexcharts';

import type { InvestmentAccount } from '@/book/models';
import Money from '@/book/Money';
import { toFixed } from '@/helpers/number';
import { currencyToSymbol } from '@/book/helpers';

// apexcharts import references window so we need this
// https://stackoverflow.com/questions/68596778
const Chart = dynamic(() => import('react-apexcharts'), { ssr: false });

export type WeightsChartProps = {
  investments: InvestmentAccount[],
  totalValue: Money,
};

const RED = '#D12B2B';
const GREEN = '#52B12C';
const LIGHT_GREEN = '#B9DFA9';
const LIGHT_RED = '#FFECEC';

function getStockColor(
  value: number,
  minValue: number,
  maxValue: number,
  minColor: string,
  maxColor: string,
): string {
  const normalizedValue = (value - minValue) / (maxValue - minValue);
  const startColor = hexToRgb(minColor);
  const endColor = hexToRgb(maxColor);
  const red = Math.round(startColor.r + normalizedValue * (endColor.r - startColor.r));
  const green = Math.round(startColor.g + normalizedValue * (endColor.g - startColor.g));
  const blue = Math.round(startColor.b + normalizedValue * (endColor.b - startColor.b));
  return rgbToHex(red, green, blue);
}

function hexToRgb(hex: string): { r: number, g: number, b: number } {
  const num = parseInt(hex.replace('#', ''), 16);
  return {
    // eslint-disable-next-line no-bitwise
    r: (num >> 16) & 255,
    // eslint-disable-next-line no-bitwise
    g: (num >> 8) & 255,
    // eslint-disable-next-line no-bitwise
    b: num & 255,
  };
}

function rgbToHex(r: number, g: number, b: number): string {
  return `#${[r, g, b].map(x => {
    const hex = x.toString(16);
    return hex.length === 1 ? `0${hex}` : hex;
  }).join('')}`;
}

export default function WeightsChart({
  investments,
  totalValue,
}: WeightsChartProps): JSX.Element {
  const currencySymbol = currencyToSymbol(totalValue.currency);
  const data: { [ticker: string]: {
    x: string,
    y: number,
    today: string,
    pct: number,
    color: string,
  } } = {};

  const percentages = investments.map(investment => investment.quoteInfo.changePct);
  const bestProfit = Math.max(...percentages);
  const worstProfit = Math.min(...percentages);

  investments.forEach(investment => {
    const investmentValueInCurrency = investment.valueInCurrency.toNumber();
    const percentage = toFixed((investmentValueInCurrency / totalValue.toNumber()) * 100);
    const todayChangePct = investment.quoteInfo.changePct;

    let color = '';
    if (todayChangePct < 0) {
      color = getStockColor(todayChangePct, worstProfit, 0, RED, LIGHT_RED);
    } else {
      color = getStockColor(todayChangePct, 0, bestProfit, LIGHT_GREEN, GREEN);
    }

    if (investmentValueInCurrency > 0) {
      data[investment.account.name] = {
        x: investment.account.name,
        y: toFixed(investmentValueInCurrency),
        today: toFixed(todayChangePct) > 0 ? `+${toFixed(todayChangePct)}%` : `${toFixed(todayChangePct)}%`,
        pct: percentage,
        color,
      };
    }
  });

  const series = [
    {
      data: Object.values(data),
    },
  ];

  const options: ApexOptions = {
    chart: {
      id: 'weightsChart',
      toolbar: {
        show: false,
      },
      sparkline: {
        enabled: true,
      },
    },
    legend: {
      show: false,
    },
    dataLabels: {
      enabled: true,
      style: {
        fontSize: '12px',
      },
      // @ts-ignore
      formatter: (text: string) => `${text}: ${data[text].today}`,
      offsetY: -4,
    },
    plotOptions: {
      treemap: {
        useFillColorAsStroke: true,
        colorScale: {
          ranges: Object.values(data).map(each => ({
            from: each.y,
            to: each.y,
            color: each.color,
          })),
        },
      },
    },
    tooltip: {
      fillSeriesColor: true,
      y: {
        formatter: (val: number, { dataPointIndex, w }) => {
          const ticker = w.globals.categoryLabels[dataPointIndex];
          return `${val}${currencySymbol} (${data[ticker].pct}%)`;
        },
      },
      theme: 'dark',
    },
  };

  return (
    <div>
      {
        (typeof window !== 'undefined')
        // https://stackoverflow.com/questions/75103994
        // @ts-ignore
        && <Chart options={options} series={series} type="treemap" width="100%" height={685} />
      }
    </div>
  );
}
