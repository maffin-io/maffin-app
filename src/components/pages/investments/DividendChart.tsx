import React from 'react';
import { Interval, DateTime } from 'luxon';
import type { ChartData } from 'chart.js';

import Money from '@/book/Money';
import Bar from '@/components/charts/Bar';
import * as API from '@/hooks/api';
import { moneyToString } from '@/helpers/number';
import type { InvestmentAccount } from '@/book/models';

export default function DividendChart(): JSX.Element {
  const [selectedYear, setSelectedYear] = React.useState(DateTime.now().year);
  let { data: investments } = API.useInvestments();
  investments = investments || [];

  const { data: currency } = API.useMainCurrency();
  const unit = currency?.mnemonic || '';

  const [yearData, monthlyData] = buildData(investments, selectedYear);

  return (
    <div>
      <span className="text-lg">
        Dividends
      </span>

      <div className="grid grid-cols-12">
        <div className="col-span-4">
          <div id="chart-year">
            <Bar
              height="400"
              data={yearData}
              options={{
                onClick: (_, elements, chart) => {
                  if (elements.length) {
                    setSelectedYear(
                      (chart.data.labels?.[elements[0].index]) as number,
                    );
                  }
                },
                maintainAspectRatio: false,
                indexAxis: 'y',
                plugins: {
                  datalabels: {
                    display: false,
                  },
                  legend: {
                    display: false,
                  },
                  tooltip: {
                    displayColors: false,
                    backgroundColor: (
                      tooltipItem,
                    ) => tooltipItem.tooltip.labelColors[0].backgroundColor,
                    callbacks: {
                      title: () => '',
                      label: (ctx) => `${moneyToString(Number(ctx.raw), unit)}`,
                    },
                  },
                },
                scales: {
                  x: {
                    grid: {
                      drawOnChartArea: false,
                    },
                    border: {
                      display: false,
                    },
                  },
                  y: {
                    border: {
                      display: false,
                    },
                    grid: {
                      display: false,
                    },
                  },
                },
              }}
            />
          </div>
        </div>

        <div className="col-span-8">
          <div id="chart-year">
            <Bar
              height="400"
              data={monthlyData}
              options={{
                maintainAspectRatio: false,
                hover: {
                  mode: 'dataset',
                  intersect: true,
                },
                scales: {
                  x: {
                    stacked: true,
                    grid: {
                      display: false,
                    },
                  },
                  y: {
                    border: {
                      display: false,
                    },
                    stacked: true,
                    ticks: {
                      maxTicksLimit: 5,
                      callback: (value) => moneyToString(value as number, unit),
                    },
                  },
                },
                plugins: {
                  datalabels: {
                    display: false,
                  },
                  legend: {
                    position: 'bottom',
                    labels: {
                      boxWidth: 12,
                    },
                  },
                  tooltip: {
                    displayColors: false,
                    backgroundColor: (
                      tooltipItem,
                    ) => tooltipItem.tooltip.labelColors[0].backgroundColor,
                    callbacks: {
                      title: () => '',
                      label: (ctx) => `${ctx.dataset.label}: ${moneyToString(Number(ctx.raw), unit)}`,
                    },
                  },
                },
              }}
            />
          </div>
        </div>
      </div>
    </div>
  );
}

function buildData(
  investments: InvestmentAccount[],
  selectedYear: number,
): [ChartData<'bar'>, ChartData<'bar'>] {
  const yearData: ChartData<'bar'> = {
    labels: [],
    datasets: [
      {
        data: [],
        backgroundColor: 'rgba(34, 197, 94, 0.2)',
        hoverBackgroundColor: 'rgba(34, 197, 94, 0.5)',
      },
    ],
  };

  const allDividends: {
    ticker: string,
    when: DateTime,
    amountInCurrency: Money,
  }[] = [];
  investments.forEach(investment => {
    investment.dividends.forEach(dividend => {
      allDividends.push({
        ticker: investment.account.name,
        ...dividend,
      });
    });
  });

  let startDate = DateTime.now();
  let endDate = DateTime.now();
  if (allDividends.length) {
    startDate = DateTime.fromMillis(
      Math.min(...(allDividends.map(dividend => dividend.when.toMillis()))),
    ) as DateTime<true>;
    endDate = DateTime.fromMillis(
      Math.max(...(allDividends.map(dividend => dividend.when.toMillis()))),
    ) as DateTime<true>;
  }

  const interval = Interval.fromDateTimes(startDate, endDate);
  const years = interval.splitBy({ year: 1 }).map(d => (d.start as DateTime).year);
  const currentYear = DateTime.now().year;
  if (!years.includes(currentYear)) {
    years.push(DateTime.now().year);
  }

  yearData.labels = years;
  yearData.datasets[0].data = new Array(years.length).fill(0);

  const monthData: ChartData<'bar'> = {
    labels: ['Jan', 'Feb', 'Mar', 'Apr', 'May', 'Jun', 'Jul', 'Aug', 'Sep', 'Oct', 'Nov', 'Dec'],
    datasets: [],
  };

  allDividends.forEach(dividend => {
    (
      yearData.datasets[0].data[years.indexOf(dividend.when.year)] as number
    ) += dividend.amountInCurrency.toNumber();

    if (dividend.when.year === selectedYear) {
      let tickerDataset = (monthData.datasets.find(dataset => dataset.label === dividend.ticker));
      if (!tickerDataset) {
        tickerDataset = { data: new Array(12).fill(0), label: dividend.ticker };
        monthData.datasets.push(tickerDataset);
      }

      (
        tickerDataset.data[dividend.when.month - 1] as number
      ) += dividend.amountInCurrency.toNumber();
    }
  });

  return [yearData, monthData];
}
