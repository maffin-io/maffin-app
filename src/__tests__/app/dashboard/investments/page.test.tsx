import React from 'react';
import { render, screen } from '@testing-library/react';
import type { SWRResponse } from 'swr';

import { Commodity } from '@/book/entities';
import InvestmentsPage from '@/app/dashboard/investments/page';
import WeightsChart from '@/components/pages/investments/WeightsChart';
import StatisticsWidget from '@/components/StatisticsWidget';
import InvestmentsTable from '@/components/pages/investments/InvestmentsTable';
import DividendChart from '@/components/pages/investments/DividendChart';
import Money from '@/book/Money';
import { InvestmentAccount } from '@/book/models';
import * as apiHook from '@/hooks/useApi';

jest.mock('@/hooks/useApi', () => ({
  __esModule: true,
  ...jest.requireActual('@/hooks/useApi'),
}));

jest.mock('@/components/pages/investments/WeightsChart', () => jest.fn(
  () => <div data-testid="WeightsChart" />,
));
const WeightsChartMock = WeightsChart as jest.MockedFunction<typeof WeightsChart>;

jest.mock('@/components/StatisticsWidget', () => jest.fn(
  () => <div data-testid="StatisticsWidget" />,
));

jest.mock('@/components/pages/investments/DividendChart', () => jest.fn(
  () => <div data-testid="DividendChart" />,
));

jest.mock('@/components/pages/investments/InvestmentsTable', () => jest.fn(
  () => <div data-testid="InvestmentsTable" />,
));

describe('InvestmentsPage', () => {
  beforeEach(() => {
    jest.spyOn(apiHook, 'default').mockReturnValue({ data: undefined } as SWRResponse);
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  it('renders while loading data', () => {
    const { container } = render(<InvestmentsPage />);

    expect(WeightsChart).toHaveBeenLastCalledWith(
      {
        investments: [],
        totalValue: expect.any(Money),
      },
      {},
    );
    expect(WeightsChartMock.mock.calls[0][0].totalValue.toString()).toEqual('0.00 EUR');
    expect(StatisticsWidget).toHaveBeenNthCalledWith(
      1,
      {
        className: 'mx-6',
        title: 'Value/Cost',
        stats: '0.00 €',
        description: '0 total invested',
      },
      {},
    );
    expect(StatisticsWidget).toHaveBeenNthCalledWith(
      2,
      {
        className: 'mx-6',
        title: 'Unrealized Profit',
        stats: '0.00 € (NaN%)',
        description: '0 (NaN%) with dividends',
        statsTextClass: '',
      },
      {},
    );
    expect(StatisticsWidget).toHaveBeenNthCalledWith(
      3,
      {
        className: 'mx-6',
        title: 'Realized',
        stats: '0.00 €',
        description: '0.00 € from dividends',
        statsTextClass: 'text-green-500',
      },
      {},
    );
    expect(DividendChart).toHaveBeenLastCalledWith(
      {
        investments: [],
      },
      {},
    );
    expect(InvestmentsTable).toHaveBeenLastCalledWith(
      {
        investments: [],
      },
      {},
    );
    expect(apiHook.default).toBeCalledTimes(2);
    expect(apiHook.default).toHaveBeenNthCalledWith(1, '/api/investments');
    expect(apiHook.default).toHaveBeenNthCalledWith(2, '/api/main-currency');
    expect(container).toMatchSnapshot();
  });

  it('renders with data', async () => {
    const mainCurrency = {
      mnemonic: 'EUR',
      guid: 'eur',
      namespace: 'CURRENCY',
    } as Commodity;

    const investment1 = {
      account: { name: 'Investment' },
      quantity: new Money(10, 'TICKER'),
      valueInCurrency: new Money(5, 'EUR'),
      costInCurrency: new Money(4, 'EUR'),
      realizedProfitInCurrency: new Money(10, 'EUR'),
      realizedDividendsInCurrency: new Money(5, 'EUR'),
    } as InvestmentAccount;

    jest.spyOn(apiHook, 'default')
      .mockReturnValueOnce({ data: [investment1] } as SWRResponse)
      .mockReturnValueOnce({ data: mainCurrency } as SWRResponse);

    const { container } = render(<InvestmentsPage />);

    await screen.findByTestId('InvestmentsTable');
    expect(WeightsChart).toHaveBeenCalledWith(
      {
        investments: [investment1],
        totalValue: expect.any(Money),
      },
      {},
    );
    expect(WeightsChartMock.mock.calls[0][0].totalValue.toString()).toEqual('5.00 EUR');
    expect(StatisticsWidget).toBeCalledTimes(3);
    expect(StatisticsWidget).toHaveBeenNthCalledWith(
      1,
      {
        className: 'mx-6',
        title: 'Value/Cost',
        stats: '5.00 €',
        description: '4 total invested',
      },
      {},
    );
    expect(StatisticsWidget).toHaveBeenNthCalledWith(
      2,
      {
        className: 'mx-6',
        title: 'Unrealized Profit',
        stats: '1.00 € (25%)',
        description: '6 (150%) with dividends',
        statsTextClass: 'text-green-500',
      },
      {},
    );
    expect(StatisticsWidget).toHaveBeenNthCalledWith(
      3,
      {
        className: 'mx-6',
        title: 'Realized',
        stats: '15.00 €',
        description: '5.00 € from dividends',
        statsTextClass: 'text-green-500',
      },
      {},
    );
    expect(DividendChart).toHaveBeenLastCalledWith(
      {
        investments: [investment1],
      },
      {},
    );
    expect(InvestmentsTable).toHaveBeenLastCalledWith(
      {
        investments: [investment1],
      },
      {},
    );
    expect(container).toMatchSnapshot();
  });

  it('renders with different mainCurrency', async () => {
    const mainCurrency = {
      mnemonic: 'USD',
      guid: 'usd',
      namespace: 'CURRENCY',
    } as Commodity;

    const investment1 = {
      account: { name: 'Investment' },
      quantity: new Money(10, 'TICKER'),
      valueInCurrency: new Money(5, 'USD'),
      costInCurrency: new Money(4, 'USD'),
      realizedProfitInCurrency: new Money(10, 'USD'),
      realizedDividendsInCurrency: new Money(5, 'USD'),
    } as InvestmentAccount;

    jest.spyOn(apiHook, 'default')
      .mockReturnValueOnce({ data: [investment1] } as SWRResponse)
      .mockReturnValueOnce({ data: mainCurrency } as SWRResponse);

    const { container } = render(<InvestmentsPage />);

    await screen.findByTestId('InvestmentsTable');
    expect(WeightsChart).toHaveBeenLastCalledWith(
      {
        investments: [investment1],
        totalValue: expect.any(Money),
      },
      {},
    );
    expect(WeightsChartMock.mock.calls[0][0].totalValue.toString()).toEqual('5.00 USD');
    expect(StatisticsWidget).toBeCalledTimes(3);
    expect(StatisticsWidget).toHaveBeenNthCalledWith(
      1,
      {
        className: 'mx-6',
        title: 'Value/Cost',
        stats: '5.00 $',
        description: '4 total invested',
      },
      {},
    );
    expect(StatisticsWidget).toHaveBeenNthCalledWith(
      2,
      {
        className: 'mx-6',
        title: 'Unrealized Profit',
        stats: '1.00 $ (25%)',
        description: '6 (150%) with dividends',
        statsTextClass: 'text-green-500',
      },
      {},
    );
    expect(StatisticsWidget).toHaveBeenNthCalledWith(
      3,
      {
        className: 'mx-6',
        title: 'Realized',
        stats: '15.00 $',
        description: '5.00 $ from dividends',
        statsTextClass: 'text-green-500',
      },
      {},
    );
    expect(DividendChart).toHaveBeenLastCalledWith(
      {
        investments: [investment1],
      },
      {},
    );
    expect(InvestmentsTable).toHaveBeenLastCalledWith(
      {
        investments: [investment1],
      },
      {},
    );
    expect(container).toMatchSnapshot();
  });

  it('doesnt pass empty positions to investments table', async () => {
    const mainCurrency = {
      mnemonic: 'EUR',
      guid: 'eur',
      namespace: 'CURRENCY',
    } as Commodity;

    const investment1 = {
      account: { name: 'Investment' },
      quantity: new Money(0, 'TICKER'),
      valueInCurrency: new Money(0, 'EUR'),
      costInCurrency: new Money(0, 'EUR'),
      realizedProfitInCurrency: new Money(10, 'EUR'),
      realizedDividendsInCurrency: new Money(5, 'EUR'),
    } as InvestmentAccount;

    jest.spyOn(apiHook, 'default')
      .mockReturnValueOnce({ data: [investment1] } as SWRResponse)
      .mockReturnValueOnce({ data: mainCurrency } as SWRResponse);

    render(<InvestmentsPage />);

    await screen.findByTestId('InvestmentsTable');
    expect(InvestmentsTable).toHaveBeenLastCalledWith(
      {
        investments: [],
      },
      {},
    );
  });
});
