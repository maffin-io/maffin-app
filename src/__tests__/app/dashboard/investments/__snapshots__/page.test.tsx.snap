// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InvestmentsPage renders with data 1`] = `
<div>
  <div
    class="header"
  >
    <span
      class="title"
    >
      Your Investments
    </span>
  </div>
  <div
    class="grid grid-cols-12 mt-4"
  >
    <div
      class="col-span-4 mt-2"
    >
      <div
        data-testid="WeightsChart"
      />
    </div>
    <div
      class="col-span-8 ml-4"
    >
      <div
        class="grid grid-cols-12"
      >
        <div
          class="col-span-4"
        >
          <div
            data-testid="StatisticsWidget"
          />
        </div>
        <div
          class="col-span-4"
        >
          <div
            data-testid="StatisticsWidget"
          />
        </div>
        <div
          class="col-span-4"
        >
          <div
            data-testid="StatisticsWidget"
          />
        </div>
      </div>
      <div
        class="card"
      >
        <div
          data-testid="DividendChart"
        />
      </div>
    </div>
  </div>
  <div
    class="py-4"
  >
    <div
      data-testid="InvestmentsTable"
    />
  </div>
</div>
`;

exports[`InvestmentsPage renders with different mainCurrency 1`] = `
<div>
  <div
    class="header"
  >
    <span
      class="title"
    >
      Your Investments
    </span>
  </div>
  <div
    class="grid grid-cols-12 mt-4"
  >
    <div
      class="col-span-4 mt-2"
    >
      <div
        data-testid="WeightsChart"
      />
    </div>
    <div
      class="col-span-8 ml-4"
    >
      <div
        class="grid grid-cols-12"
      >
        <div
          class="col-span-4"
        >
          <div
            data-testid="StatisticsWidget"
          />
        </div>
        <div
          class="col-span-4"
        >
          <div
            data-testid="StatisticsWidget"
          />
        </div>
        <div
          class="col-span-4"
        >
          <div
            data-testid="StatisticsWidget"
          />
        </div>
      </div>
      <div
        class="card"
      >
        <div
          data-testid="DividendChart"
        />
      </div>
    </div>
  </div>
  <div
    class="py-4"
  >
    <div
      data-testid="InvestmentsTable"
    />
  </div>
</div>
`;
